exec-c-api/src/capi_memory.rs

```diff
@@ -0,0 +1,65 @@
+//! Instantiate a module, call functions, and read exports.
+
+use crate::{
+    capi_instance::{vm_exec_instance_t, CapiInstance},
+    service_singleton::with_service,
+    vm_exec_result_t,
+};
+use std::ptr;
+
+/// Gets the size in bytes of the memory data.
+///
+/// This function returns 0 if `memory` is a null pointer.
+#[allow(clippy::cast_ptr_alignment)]
+#[no_mangle]
+pub unsafe extern "C" fn vm_exec_instance_memory_data_length(
+    instance_ptr: *mut vm_exec_instance_t,
```

Comment:
we should mark this as `const*`, also we can use `cast_input_const_ptr!` macro from metering pr after merge